In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

---
# Loading the `nycflights13` datasets
This dataset contains information on all the flights departed from New York City in 2013.

The [Python nycflights13](https://pypi.org/project/nycflights13/) data package provides the same data as the [R nycflights13](https://cran.r-project.org/web/packages/nycflights13/index.html) package.

In [2]:
# install the package
!pip install nycflights13

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 41.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for nycflights13: filename=nycflights13-0.0.3-py3-none-any.whl size=8732740 sha256=c58d5b6233db38ee4365541f1d425dbd84e6a50e214f51bafa0393abbae49ef3
  Stored in directory: /root/.cache/pip/wheels/0e/b7/7b/c129c6a2717d8825caa178f3b07e260cfb12c39f95fd165ff1
Successfully built nycflights13


In [3]:
# load the `flights` table
from nycflights13 import flights
print(type(flights))
flights.head()

<class 'pandas.core.frame.DataFrame'>


year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
2      1089     5      40  2013-01-01T10:00:00Z  
3      1576     5      45  2013-01-01T10:00:00Z  
4       762     6       0  2013-01-01T11:00:00Z

In [4]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   arr_time        328063 non-null  float64
 7   sched_arr_time  336776 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         336776 non-null  object 
 10  flight          336776 non-null  int64  
 11  tailnum         334264 non-null  object 
 12  origin          336776 non-null  object 
 13  dest            336776 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        336776 non-null  int64  
 16  hour            336776 non-null  int64  
 17  minute    

In [5]:
# load the `weather` table
from nycflights13 import weather
print(type(weather))
weather.head()

<class 'pandas.core.frame.DataFrame'>


origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

In [6]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26115 entries, 0 to 26114
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   origin      26115 non-null  object 
 1   year        26115 non-null  int64  
 2   month       26115 non-null  int64  
 3   day         26115 non-null  int64  
 4   hour        26115 non-null  int64  
 5   temp        26114 non-null  float64
 6   dewp        26114 non-null  float64
 7   humid       26114 non-null  float64
 8   wind_dir    25655 non-null  float64
 9   wind_speed  26111 non-null  float64
 10  wind_gust   5337 non-null   float64
 11  precip      26115 non-null  float64
 12  pressure    23386 non-null  float64
 13  visib       26115 non-null  float64
 14  time_hour   26115 non-null  object 
dtypes: float64(9), int64(4), object(2)
memory usage: 3.0+ MB


---
# &diams; Predict daily departure delays using weather information

Suppose we are interested in using both `flights` and `weather` data to build a regression model to predict flight departure delays using the weather information.

---

**Task-1** The `weather` data contains **hourly** meteorological data for LGA, JFK and EWR.

Calculate the **daily** weather information at **LGA** in 2013. Here we only consider the following six weather-related variables: 1) temperature, 2) humidity, 3) wind speed, 4) precipitation, 5) sea level pressure, and 6) visibility.

Save your results in a pandas DataFrame called `weather_LGA_daily`. The DataFrame should have nine columns: year, month, day, followed by the six daily average weather variables.

Expected results: The resulting table is of shape 364 x 9; the first row should be `2013	1	1	37.227826	53.883478	15.410445	0.00	1012.622727	9.956522`.

**Note**: It helps to use `reset_index()` after grouping so the grouped result won't have a hierarchical index;

In [7]:
#ANSWER
# Filtering the data for LGA and 2013
weather_LGA = weather[(weather.origin == 'LGA') & (weather.year == 2013)]

# Grouping data by year, month, and day using groupby and calculating means
weather_LGA_daily = weather_LGA.groupby(['year', 'month', 'day']).agg({
    'temp': np.nanmean,
    'humid': np.nanmean,
    'wind_speed': np.nanmean,
    'precip': np.nanmean,
    'pressure': np.nanmean,
    'visib': np.nanmean
}).reset_index()

# To show first row of the resulting DataFrame
print(weather_LGA_daily)


     year  month  day       temp      humid  wind_speed  precip     pressure  \
0    2013      1    1  37.227826  53.883478   15.410445    0.00  1012.622727   
1    2013      1    2  28.790000  50.195000   13.617563    0.00  1018.062500   
2    2013      1    3  30.267500  52.682500   11.076257    0.00  1020.320833   
3    2013      1    4  35.832500  49.893333   15.391682    0.00  1016.441667   
4    2013      1    5  38.277500  46.420417    9.877528    0.00  1022.458333   
..    ...    ...  ...        ...        ...         ...     ...          ...   
359  2013     12   26  33.335000  58.893750    8.343155    0.00  1025.250000   
360  2013     12   27  36.462500  47.760833   11.507800    0.00  1026.475000   
361  2013     12   28  44.967500  42.286250   10.740613    0.00  1021.316667   
362  2013     12   29  43.857500  67.425833    9.829579    0.05  1011.538889   
363  2013     12   30  39.370526  62.662105   12.779715    0.00  1013.282353   

         visib  
0     9.956522  
1    

---
**Task-2 : Missing Value Treatment** The `flights` data contains all 336,776 flights
that departed from New York City in 2013. 

There are some missing values in the flights data set which were caused by canceled flights. We can clean the flights data by removing flight record that has missing values in `dep_delay` or `arr_delay`, and save the non-canceled flights in a new pandas DataFrame `not_canceled`. 

The resulting `not_canceled` table should have 327,346 rows.

We can cross-check using pandas functions `isna()`, `notna()`, and `dropna()`.

In [8]:
#ANSWER
# Removing the rows with missing values in dep_delay or arr_delay columns
not_canceled = flights.dropna(subset=['dep_delay', 'arr_delay'])

# To see the number of rows in the resulting table
print(not_canceled)


        year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336765  2013      9   30    2240.0            2245       -5.0    2334.0   
336766  2013      9   30    2240.0            2250      -10.0    2347.0   
336767  2013      9   30    2241.0            2246       -5.0    2345.0   
336768  2013      9   30    2307.0            2255       12.0    2359.0   
336769  2013      9   30    2349.0            2359      -10.0     325.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819    

---
**Task-3** Next, use the `not_canceled` flights data to calculate the **daily** average departure delays at **LGA** in 2013. Save your results as `LGA_delay_daily`. The DataFrame should have four columns: year, month, day,
and the daily average departure delay. 

What is the largest and smallest daily average departure delay?
How many days in 2013 has negative daily average departure delays?

Expected results: The resulting table is of shape 365 x 4; the first row should be `2013	1	1	3.203390`; the max and min values are 106.210526 and -3.243243; 56 days have negative daily average departure delays.

In [9]:
#ANSWER
# Filtering for LGA flights having only the year, month, day, & dep_delay columns
not_canceled_LGA = not_canceled[not_canceled['origin'] == 'LGA'][['year', 'month', 'day', 'dep_delay']]

#Grouping by year, month, & day then calculating the mean of the dep_delay column
LGA_delay_daily = not_canceled_LGA.groupby(['year', 'month', 'day']).agg({'dep_delay': 'mean'}).reset_index()

# Printing the result and summary statistics
print(LGA_delay_daily)
print('Max daily average departure delay:', LGA_delay_daily['dep_delay'].max())
print('Min daily average departure delay:', LGA_delay_daily['dep_delay'].min())
print('Number of days with negative daily average departure delays:', sum(LGA_delay_daily['dep_delay'] < 0))

     year  month  day  dep_delay
0    2013      1    1   3.203390
1    2013      1    2   5.633333
2    2013      1    3  10.825397
3    2013      1    4   2.929412
4    2013      1    5   2.450000
..    ...    ...  ...        ...
360  2013     12   27   6.581395
361  2013     12   28   3.013100
362  2013     12   29   9.650376
363  2013     12   30   4.855705
364  2013     12   31   2.631818

[365 rows x 4 columns]
Max daily average departure delay: 106.21052631578948
Min daily average departure delay: -3.2432432432432434
Number of days with negative daily average departure delays: 56


---
**Task-4**: Now join the two tables `weather_LGA_daily` and `LGA_delay_daily` that you obtained from parts (i) and (iii), save the result as `LGA_delay_weather_daily`. 

The resulting table should be of shape 364 x 10.


In [10]:
#ANSWER
#joining the two tables
LGA_delay_weather_daily = pd.merge(weather_LGA_daily, LGA_delay_daily, on=['year', 'month', 'day'])
LGA_delay_weather_daily



year  month  day       temp      humid  wind_speed  precip     pressure  \
0    2013      1    1  37.227826  53.883478   15.410445    0.00  1012.622727   
1    2013      1    2  28.790000  50.195000   13.617563    0.00  1018.062500   
2    2013      1    3  30.267500  52.682500   11.076257    0.00  1020.320833   
3    2013      1    4  35.832500  49.893333   15.391682    0.00  1016.441667   
4    2013      1    5  38.277500  46.420417    9.877528    0.00  1022.458333   
..    ...    ...  ...        ...        ...         ...     ...          ...   
359  2013     12   26  33.335000  58.893750    8.343155    0.00  1025.250000   
360  2013     12   27  36.462500  47.760833   11.507800    0.00  1026.475000   
361  2013     12   28  44.967500  42.286250   10.740613    0.00  1021.316667   
362  2013     12   29  43.857500  67.425833    9.829579    0.05  1011.538889   
363  2013     12   30  39.370526  62.662105   12.779715    0.00  1013.282353   

         visib  dep_delay  
0     9.956522   3.203390  
1    10.000000   5.633333  
2    10.000000  10.825397  
3    10.000000   2.929412  
4    10.000000   2.450000  
..         ...        ...  
359  10.000000  11.630000  
360  10.000000   6.581395  
361  10.000000   3.013100  
362   7.593750   9.650376  
363  10.000000   4.855705  

[364 rows x 10 columns]

---

**Task-5** Use the sklearn `model_selection.train_test_split()` function to separate the `LGA_delay_weather_daily` data 70/30 into a training set and a testing set. Use random seed 206.

In [11]:
#ANSWER
from sklearn.model_selection import train_test_split

X = LGA_delay_weather_daily.drop(columns=['dep_delay'])
y = LGA_delay_weather_daily['dep_delay']
#separating the LGA_delay_weather_daily data 70/30 into a training set and a testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=206)


---
**Task-5** Suppose we are interested in applying a multiple linear regression model to to predict the daily average departure delays at LGA, using the following six weather predictors: 1) temperature, 2) humidity, 3) wind speed, 4) precipitation, 5) sea level pressure, and 6) visibility.

Fit your model on the training data, and print out the coefficients.

Expected results: the coefficients are `[ 1.55083493e-01  2.97484369e-01  3.09784108e-01  1.71880854e+02  -1.51085105e-01 -9.80963451e-01] 142.29623230752145`



In [12]:
#ANSWER
from sklearn.linear_model import LinearRegression

# Selecting the weather predictors
X_train_weather = X_train[['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']]

# LinearRegression class is being created
reg = LinearRegression()

# Fitting the model on the training dataset
reg.fit(X_train_weather, y_train)

# Coefficients of the model
coefficients = np.append(reg.coef_, reg.intercept_)

# Showing coefficients
print(coefficients)


# printing the output
print(reg.intercept_)



[ 1.55083493e-01  2.97484369e-01  3.09784108e-01  1.71880854e+02
 -1.51085105e-01 -9.80963451e-01  1.42296232e+02]
142.29623230752145


---
**Task-7** Calculate the root mean squared error (RMSE) and Rˆ2 metrics on the training data.

Expected results: 10.68017531753989 and 0.3382017700023695

**Hint:** Check sklearn `metrics` subpackage.

In [13]:
#ANSWER
from sklearn.metrics import mean_squared_error, r2_score

# Prediction of the departure delays by the training data
y_train_pred = reg.predict(X_train_weather)

# Calculating the RMSE
rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))

# Calculating the R^2 score
r2 = r2_score(y_train, y_train_pred)

# Printing the RMSE & R^2 score
print("RMSE:", rmse)
print("R^2:", r2)



RMSE: 10.68017531753989
R^2: 0.3382017700023695


---

**Task-8**  Now apply the trained model on the testing data, and report the RMSE and Rˆ2 metrics.

Expected results: 14.31635802661413 and 0.31131710327784845

In [14]:
# ANSWER
# Selecting weather predictors
X_test_weather = X_test[['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']]

# Predicting the departure delays
y_test_pred = reg.predict(X_test_weather)

#Calculating the RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

# Calculating the R^2 score
r2 = r2_score(y_test, y_test_pred)

# Printing the RMSE and R^2 score
print("RMSE:", rmse)
print("R^2:", r2)


RMSE: 14.31635802661413
R^2: 0.31131710327784845


---
**Task-9** Calculate the 5-fold cross-validation RMSE and R^2 of your linear regression model.

**Resources**: (1) Check the `sklearn.model_selection.cross_val_score()` function; 
(2) Pay attention to the `scoring` parameter; the predefined scoring metrics are available at https://scikit-learn.org/stable/modules/model_evaluation.html#the-scoring-parameter-defining-model-evaluation-rules


In [15]:
#ANSWER
from sklearn.model_selection import cross_val_score

# Fistly, selecting weather predictors from the data
X_weather = X[['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']]

# Then, calculating the cross-validation RMSE & R^2
cv_rmse = np.sqrt(-cross_val_score(reg, X_weather, y, cv=5, scoring='neg_mean_squared_error'))
cv_r2 = cross_val_score(reg, X_weather, y, cv=5, scoring='r2')

# Printing the average RMSE and R^2
print("Cross-validation RMSE:", cv_rmse.mean())
print("Cross-validation R^2:", cv_r2.mean())



Cross-validation RMSE: 12.75963240180539
Cross-validation R^2: 0.1657775814293076


---
**Task-10** The performance of the current linear regression model doesn’t look very good. Please propose another regression to predict the the daily average departure delays at LGA. You can use either a different set of predictors, or a completely new regression model.

Calculate the 5-fold cross-validation RMSE and R^2 metrics.



In [ ]:
#ANSWER
from sklearn.ensemble import RandomForestRegressor
from sklearn.impute import SimpleImputer

# Modifing the DataFrame
LGA_delay_weather_daily.dropna(inplace=True)

#fill in any missing values in the training and testing data
imputer = SimpleImputer(strategy='mean')
X_train = imputer.fit_transform(X_train)
X_test = imputer.transform(X_test)

# Defining the predictors & target variable
predictors = ['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']
target = 'dep_delay'

# Splitting the data into training & testing sets
X_train, X_test, y_train, y_test = train_test_split(LGA_delay_weather_daily[predictors],LGA_delay_weather_daily[ target ], test_size=0.3, random_state=206)

# Training the Random Forest Regressor model
rf = RandomForestRegressor(n_estimators=100, random_state=206)
rf.fit(X_train, y_train)

# Calculating the 5-fold cross-validation RMSE and R^2 metrics
cv_rmse = np.sqrt(-cross_val_score(rf, LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], cv=5, scoring='neg_mean_squared_error'))
cv_r2 = cross_val_score(rf, LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], cv=5, scoring='r2')

# Printing the results
print(f"5-fold CV RMSE: {cv_rmse.mean()} +/- {cv_rmse.std()}")
print(f"5-fold CV R^2: {cv_r2.mean()} +/- {cv_r2.std()}")

# Making predictions on the testing data and calculate the RMSE and R^2 metrics
y_pred = rf.predict(X_test)
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
test_r2 = r2_score(y_test, y_pred)
print(f"Test RMSE: {test_rmse}")
print(f"Test R^2: {test_r2}")


---
# Classify daily departure delays using weather information [40 pts]

In this question, we attempt to reduce the problem to a classification task.

---
**Task-1** First, label each day into two status groups: `delayed` if the daily average departure delay is greater than 0; `not-delayed`, otherwise.
Add a new column `delay_label` to `LGA_delay_weather_daily` to present this binary label. How many days are labeled as `delayed`? And `not-delayed`?

The expected answers are 308 days as `delayed` and 56 days as `not-delayed`.


In [ ]:
#ANSWER
# Creating a new column named "delay_label" based on the daily average departure delay
LGA_delay_weather_daily["delay_label"]= np.where(LGA_delay_weather_daily["dep_delay"] > 0, "delayed", "not-delayed")

# Counting the number of days labeled as delayed or not-delayed
num_delayed = LGA_delay_weather_daily[LGA_delay_weather_daily["delay_label"] == "delayed"].shape[0]
num_not_delayed = LGA_delay_weather_daily[LGA_delay_weather_daily["delay_label"] == "not-delayed"].shape[0]

# Printing the number of days labeled as delayed or not-delayed
print(f"Number of delayed days: {num_delayed}")
print(f"Number of not-delayed days: {num_not_delayed}")

---
**Task-2** Similarly, use the `sklearn.model_selection.train_test_split()` function to separate the LGA_delay_weather_daily data 70/30 into a training set and a testing set. Use random seed 206.

In [ ]:
#ANSWER
from sklearn.model_selection import train_test_split

# Defining the predictors & target variable
predictors = ['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']
target= 'delay_label'

# Splitting the data into training and testing sets(70/30) 
X_train, X_test, y_train, y_test = train_test_split(LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], test_size=0.3, random_state=206)


---
**Task-2** After task-2, apply a K-nearest-neighbor model (K=3) to classify the daily average departure delay status at LGA, i.e. `delay_label`, using the six weather predictors: 1) temperature, 2) humidity, 3) wind speed, 4) precipitation, 5) sea level pressure, and 6) visibility.

Fit the model on the training data, what is the misclassification error rate (i.e., 1 - accuracy) on the training data?

The expected result is 0.10236220472440949.

In [ ]:
# ANSWER
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Defining the predictors and target variable
predictors = ['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']
target = 'delay_label'

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], test_size=0.3, random_state=206)

# Training the KNN model with K=3
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train, y_train)

# Calculating the misclassification error rate on the training data
y_pred_train = knn.predict(X_train)
train_error = 1 - accuracy_score(y_train, y_pred_train)
print(f"Training Misclassification Error Rate: {train_error}")

---
**Task-4** What is the misclassification error rate on the test data?

The expected result is 0.1636363636363637.

In [ ]:
# ANSWER
from sklearn.metrics import accuracy_score

# Making predictions on the test data
y_pred_test = knn.predict(X_test)

# Calculating the misclassification error rate on the test data
test_error_rate= 1 - accuracy_score(y_test, y_pred_test)

print(f"Test Misclassification Error Rate: {test_error_rate:}")



---
**Task-5** Print out the confusion matrix for the classification results on the test data.

Calculate the false positive rate (FPR), true positive rate (TPR), and false discovery rate (FDR).


In [ ]:
# ANSWER
from sklearn.metrics import confusion_matrix

# Predictions on the test data using trained KNN model
y_pred = knn.predict(X_test)

# Calculating the confusion matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion matrix:")
print(cm)

#Calculating the false positive rate (FPR), true positive rate (TPR), & false discovery rate (FDR)
tn, fp, fn, tp = cm.ravel()
fpr = fp / (fp + tn)
tpr = tp / (tp + fn)
fdr = fp / (fp + tp)
print(f"False positive rate (FPR): {fpr}")
print(f"True positive rate (TPR): {tpr}")
print(f"False discovery rate (FDR): {fdr}")


---
**Task-6** Plot the ROC curve of this KNN classifier.


In [ ]:
# ANSWER
from sklearn.metrics import roc_curve, RocCurveDisplay

#converting the target variable y_test into a binary variable
y_test_binary = (y_test == 'not-delayed').astype(int)

# Making predictions on the testing data using the trained KNN model
y_pred_proba = knn.predict_proba(X_test)[:,1]

# Calculating the ROC curve
fpr, tpr, thresholds = roc_curve(y_test_binary, y_pred_proba)


# Plotting the ROC curve
roc_display = RocCurveDisplay(tpr=tpr,fpr=fpr, roc_auc=None).plot()
roc_display.ax_.set_title('ROC Curve for KNN Classifier: Positive Label is (Not-delayed) for both axis')
roc_display.ax_.set_xlabel('False Positive Rate (FPR)')
roc_display.ax_.set_ylabel('True Positive Rate (TPR)')


---
**Task-7** Calculate the 5-fold cross-validation misclassification rate of your KNN model.


In [ ]:
# ANSWER
from sklearn.model_selection import cross_val_score

# Defining the predictors and target variable
predictors = ['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']
target = 'delay_label'

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], test_size=0.3, random_state=206)

# Training the KNN model with K=3
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train,y_train)

# Cross-validation with 5 folds
scores = cross_val_score(knn, LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], cv=5)

# Misclassification rate (1 - accuracy)
misclassification_rate = 1 - scores.mean()

print(f"Cross-validation misclassification rate: {misclassification_rate:.3f}")


---
**Task-8** Construct another classifier to predict the binary `delayed`/`not-delayed` status.

You can use either a different selection of predictors, or a different classification model. Calculate the 5-fold cross-validation misclassification error rate of your classifier.


**Answer** I used many classifiers such as Logistic regression, Random Forest classifier, decision tree classifier, KNN, Support vector machines classifier (SVC). I found that SVC works the best among these. For simplicity, I'm just showing SVC results.  

I also tried deep learning based methods such as Convolution nueral networks (CNN), RNN, and MLP(Multi-Layer Processing) but these methods require the inputs to be in specific format. Therefore, I need to reshape my data in order to be a compatible input for these methods. I'm not sure if after reshaping out data will have the same meaning. Therefore, conceptually I don't want to use these methods in the context of our problem. 

Overall, my selected classifier SVC() gives the classification error to be 15.38%. Simple logistic classifier gives us more than 16% error. 

I don't know whether I'll get the extra credit or not, but this question was awesome. I learned many techniques and their comparative statistics.

In [ ]:
# ANSWER
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score

# Defining the predictors and target variable
predictors = ['temp', 'humid', 'wind_speed', 'precip', 'pressure', 'visib']
target = 'delay_label'

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], test_size=0.3, random_state=206)

# Creating and fitting a classifier : Support vector regressors
clf = SVC()
clf.fit(X_train, y_train)

# Making predictions on the testing data
y_pred = clf.predict(X_test)

# Calculating the accuracy score of the logistic regression model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy of Support Vector Classifier model:",accuracy)

# Calculating the 5-fold cross-validation misclassification rate
cv_scores = cross_val_score(clf, LGA_delay_weather_daily[predictors], LGA_delay_weather_daily[target], cv=5)
misclassification_rate = 1 - cv_scores.mean()
print("5-fold cross-validation misclassification rate of SVC model:", misclassification_rate)
